File: 백준/Silver/9012. 괄호/괄호.java
import java.util.*;
import java.io.*;

public class Main {
    private final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private final BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    private final StringBuilder sb = new StringBuilder();
    private StringTokenizer st;

    private void solution() throws Exception {
        int n = Integer.parseInt(br.readLine());

        for (int i = 0; i < n; i++) {
            sb.append((isVPS(br.readLine()) ? "YES" : "NO") + "\n");
        }

        bw.write(sb + "");
        bw.flush();
        bw.close();
    }

    private boolean isVPS(String data) {
        int n = data.length();
        Deque<Character> stack = new ArrayDeque<>();

        for (int i = 0; i < n; i++) {
            if (data.charAt(i) == '(') {
                stack.add('(');
            } else {
                if (stack.isEmpty()) {
                    return false;
                }
                if (stack.peekLast() != '(') {
                    return false;
                }
                stack.pollLast();
            }
        }

        return stack.isEmpty();
    }

    public static void main(String[] args) throws Exception {
        new Main().solution();
    }
}

File: 백준/Silver/9012. 괄호/괄호.py
from collections import deque
import sys

input = sys.stdin.readline


def is_vps(data):
    stack = deque()

    for x in data:
        if x == '(':
            stack.append('(')
        else:
            if len(stack) == 0:
                return False
            if stack[-1] != '(':
                return False
            stack.pop()

    return len(stack) == 0


n = int(input())

for _ in range(n):
    if is_vps(input().strip()):
        print("YES")
    else:
        print("NO")
